### PR TITLE
WIP: Integrate Federation Into Schema Builder

### PR DESCRIPTION
**Pull request by shmulik-klein**
Initial PR for #521

### CODE REVIEW

File: common/schema-builder/pom.xml [hunk 1]
```diff
@@ -48,6 +48,11 @@
             <artifactId>smallrye-graphql-api</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>io.smallrye</groupId>
+            <artifactId>smallrye-graphql-federation-api</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (t1) [inline common/schema-builder/pom.xml:55]: I think we should move all the federation annotations directly into the api module, maybe in a separate package. @phillip-kruger: what do you think?

File: common/schema-builder/src/test/java/io/smallrye/graphql/index/SchemaBuilderTest.java [hunk 2]
```diff
@@ -133,6 +133,43 @@ public void testSchemaWithDuplicates() {
...
+            Indexer indexer = new Indexer();
+            indexDirectory(indexer, "io/smallrye/graphql/index/app");
+            IndexView index = indexer.complete();
+
+            final Schema schema = SchemaBuilder.build(index);
+
+            Assertions.assertFalse(schema.isFederated());
+        } catch (SchemaBuilderException e) {
+            // ok
+        }
```
Comment (t1) [inline common/schema-builder/src/test/java/io/smallrye/graphql/index/SchemaBuilderTest.java:151]: why is this okay?

File: common/schema-builder/src/test/java/io/smallrye/graphql/index/SchemaBuilderTest.java [hunk 3]
```diff
@@ -133,6 +133,43 @@ public void testSchemaWithDuplicates() {
...
+    @Test
+    public void testSchemaWithoutFederatedDirectives() {
+        try {
+            Indexer indexer = new Indexer();
+            indexDirectory(indexer, "io/smallrye/graphql/index/app");
+            IndexView index = indexer.complete();
+
+            final Schema schema = SchemaBuilder.build(index);
+
+            Assertions.assertFalse(schema.isFederated());
```
Comment (t1) [inline common/schema-builder/src/test/java/io/smallrye/graphql/index/SchemaBuilderTest.java:148]: We generally use a static import for the Assertions.

File: common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Schema.java [hunk 4]
```diff
@@ -33,7 +34,13 @@
...
     private Map<String, ErrorInfo> errors = new HashMap<>();
 
-    public Schema() {
+
+    /**
+     * Constructor
+     * @param isFederated indicates if the schema is a federated schema
+     */
+    public Schema(boolean isFederated) {
+        this.isFederated = isFederated;
```
Comment (t1) [inline common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Schema.java:43]: All the other schema details are set with setters, not with a constructor parameter. Just for uniformity, we should stick to that, or is there a good reason to change it?

File: common/schema-builder/pom.xml (hunk 1 above)
Comment (phillip-kruger) [reply common/schema-builder/pom.xml:55]: Yes this should move to the api module

File: common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Schema.java (hunk 4 above)
Comment (phillip-kruger) [reply common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Schema.java:43]: Yes for Quarkus we need a public constructor and getters and setters

File: common/schema-builder/pom.xml (hunk 1 above)
Comment (shmulik-klein) [reply common/schema-builder/pom.xml]: Copied into the seperated `server/api/src/main/java/io/smallrye/graphql/api/federation` package. 
Please let me know if you want it to be in the root package together with the core Graph QL annotations.

**Comment by t1** [comment]
Sorry for not coming back to this earlier, I was on vacation. How far do you think you are? I'm just exploring the code a bit.

File: common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Schema.java [hunk 5]
```diff
@@ -33,6 +34,10 @@
 
     private Map<String, ErrorInfo> errors = new HashMap<>();
 
+    /**
+     * Constructor
+     *
+     */
```
Comment (t1) [inline common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Schema.java:40]: don't comment the obvious ;-)
```suggestion
```

**Comment by shmulik-klein** [comment]
> Sorry for not coming back to this earlier, I was on vacation. How far do you think you are? I'm just exploring the code a bit.

Sorry for the late reply.

Not too far, I was trying to replace the event listening on `io/smallrye/graphql/federation/impl/Federation.java:111` into an explicit call to generate the federated schema before the schema is build on `io/smallrye/graphql/bootstrap/Bootstrap.java:174`.

**Comment by t1** [comment]
@shmulik-klein: the federation topic is getting more traction in my current project. Can you find some time to work on this in the near future? I completely understand how limited time is.

**Comment by robp94** [comment]
@t1 I added another pr at https://github.com/t1/graphql-federation-demo/pull/6.

There is still a problem left.
On start up, there is a crash at https://github.com/smallrye/smallrye-graphql/blob/12ea1257a46c68c90851b59eaa53b92ce22340b2/server/federation/runtime/src/main/java/io/smallrye/graphql/federation/impl/Federation.java#L144
The problem is the line above
https://github.com/smallrye/smallrye-graphql/blob/12ea1257a46c68c90851b59eaa53b92ce22340b2/server/federation/runtime/src/main/java/io/smallrye/graphql/federation/impl/Federation.java#L143
Without it, the war demo does work again.
Do we need this for something, or could we remove it?

For Quarkus there is a second problem (maybe more). The Federation.class needs the jandex index. The index is created in the quarkus extension
https://github.com/quarkusio/quarkus/blob/e24742a2f3da1572e7529dfd67b36f90abc7f704/extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java#L209
 and stored in the ScanningContext  But when it is needed it is not there anymore since we have a different current. Do you maybe have an idea how to fix this, or who could have an idea?

**Comment by johgoe** [comment]
@ robp94 Just an idea: 
if https://github.com/quarkusio/quarkus/issues/20802 is fixed

a configuration like 
```
quarkus.smallrye-graphql.events.enabled=true
quarkus.smallrye-graphql.schema-include-scalars=true
quarkus.index-dependency.graphql-federation-runtime.group-id=io.smallrye
quarkus.index-dependency.graphql-federation-runtime.artifact-id=smallrye-graphql-federation-api
quarkus.index-dependency.graphql-federation-api.group-id=io.smallrye
quarkus.index-dependency.graphql-federation-api.artifact-id=smallrye-graphql-federation-api
```

and maven dependencies like
```
      <dependency>
        <groupId>io.smallrye</groupId>
        <artifactId>smallrye-graphql-federation-api</artifactId>
        <version>1.3.4</version>
      </dependency>
      <dependency>
        <groupId>io.smallrye</groupId>
        <artifactId>smallrye-graphql-federation-runtime</artifactId>
        <version>1.3.4</version>
      </dependency>
```
should fix the quarkus problems on startup.

**Comment by phillip-kruger** [comment]
@t1 - this looks more like what we need to do - update the model to contain info on federation, and use that info in Bootstrap

**Comment by robp94** [comment]
@jmartisk @phillip-kruger since federation is now supported I think this could be closed.

**Pull request closed without merging**